tempesta_fw/http_msg.c

```diff
@@ -474,6 +476,52 @@ __hdr_del(TfwHttpMsg *hm, int hid)
 }
 
 /**
+ * Optimization that recognizes when duplicate headers immediately follows the
+ * original, and then uses the space of such a headers for the substitution.
+ * Also, it is required that all the headers must reside in one SKB.
+ * Mainly intended for multiple "Set-Cookie:" headers.
+ */
+static inline int
+__hdr_sub_dup(TfwHttpMsg *hm, TfwStr *orig_hdr, const TfwStr *hdr, int hid)
+{
+	unsigned long len = 0;
+	const TfwStr *s1 = TFW_STR_CHUNK(orig_hdr, 0);
+	const TfwStr *s2 = TFW_STR_LAST(orig_hdr);
+	const TfwStr *c1 = TFW_STR_CHUNK(s1, 0);
+	const TfwStr *c2 = TFW_STR_LAST(s2);
+	TfwStr tail = {};
+
+	if (unlikely(c1->skb != c2->skb))
+		goto slow_path;
```

Comment:
I think this is fine. Yes, probably we can adjust HTTP parser to track continuous duplicates. The adjustment isn't so trivial since some of N duplicates can be continuous and some of them aren't, so just flags aren't enough. HTTP parser is a hot spot, so generally it's unwished to complicate it to get more performance in other, not so hot, place.  Meantime, we can assume that HTTP headers, including probably duplicate Set-Cookies, are usually fit 64KB, one hardware GROed skb, so it's expectable that the condition is usually met.